storage/src/certificate_store.rs

```diff
@@ -0,0 +1,603 @@
+        // The key is basically a composite of the dictated round and
+        // the possible smallest value of the certificate digest (all byte values
+        // should be zero).
+        let key = (round, CertificateDigest::default());
+
+        let mut certificates = Vec::new();
+        for (key, _) in self.certificate_ids_by_round.iter().skip_to(&key)? {
+            let certificate = self.certificates_by_id.get(&key.1)?.ok_or_else(|| {
+                RocksDBError(format!(
+                    "Certificate with id {} not found in main storage although it should",
+                    key.1
+                ))
+            })?;
+
+            certificates.push(certificate);
+        }
+
+        Ok(certificates)
+    }
+
+    /// Retrieves the certificates of the last round
+    pub fn last_round(&self) -> StoreResult<Vec<Certificate>> {
+        // starting from the last element - hence the last round - move backwards until
+        // we find certificates of different round.
+        let mut last = self.certificate_ids_by_round.iter().skip_to_last();
+
+        let mut round = 0;
+        let mut certificates = Vec::new();
+
+        while let Some((key, _value)) = last.prev() {
+            let (certificate_round, certificate_id) = key;
+
+            // We treat zero as special value (round unset) in order to
+            // capture the last certificate's round.
+            if round == 0 {
+                round = certificate_round;
+            }
+
+            // We are now in a different round so we want to
+            // stop consuming anymore
+            if round != certificate_round {
+                break;
+            }
+
+            let certificate = self
+                .certificates_by_id
+                .get(&certificate_id)?
+                .ok_or_else(|| {
+                    RocksDBError(format!(
+                        "Certificate with id {} not found in main storage although it should",
+                        certificate_id
+                    ))
+                })?;
+
+            certificates.push(certificate);
+        }
+
+        Ok(certificates)
+    }
+
+    /// Clears both the main storage of the certificates and the secondary index
+    pub fn clear(&self) -> StoreResult<()> {
+        self.certificates_by_id.clear()?;
+        self.certificate_ids_by_round.clear()
+    }
+
+    /// Checks whether the storage is empty. The main storage is
+    /// being used to determine this.
+    pub fn is_empty(&self) -> bool {
+        self.certificates_by_id.is_empty()
+    }
+
+    /// Notifies the subscribed ones that listen on updates for the
+    /// certificate with the provided id. The obligations are notified
+    /// with the provided value. The obligation entries under the certificate id
+    /// are removed completely. If we fail to notify an obligation we don't
+    /// fail and we rather print a warn message.
+    fn notify_subscribers(&self, id: CertificateDigest, value: Certificate) {
```

Comment:
We might want to notify the subscribers in an async way - ex spin up a tokio task (probably used a bound executor / task pool for that) to notify them so we make sure we don't block the write/read path. Right now I don't consider it a huge issue as sending to the `oneshot` channel is non blocking and under normal circumstances we don't expect to have that many `notify_read` for the same values - even if we do I believe we can handle ok.

Reply:
Yes, that seems reasonable to me.